File: convert-tsx-to-jsx.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Protected files that should not be renamed
const PROTECTED_FILES = [
  'ImageWithFallback.tsx'
];

function getAllTsxFiles(dir, fileList = []) {
  const files = fs.readdirSync(dir);
  
  files.forEach(file => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    
    if (stat.isDirectory()) {
      getAllTsxFiles(filePath, fileList);
    } else if (file.endsWith('.tsx') && !PROTECTED_FILES.includes(file)) {
      fileList.push(filePath);
    }
  });
  
  return fileList;
}

function renameFile(oldPath) {
  const newPath = oldPath.replace(/\.tsx$/, '.jsx');
  try {
    fs.renameSync(oldPath, newPath);
    console.log(`✓ Renamed: ${oldPath} -> ${newPath}`);
    return true;
  } catch (error) {
    console.error(`✗ Error renaming ${oldPath}:`, error.message);
    return false;
  }
}

function updateFileImports(filePath) {
  try {
    let content = fs.readFileSync(filePath, 'utf8');
    const originalContent = content;
    
    // Update .tsx imports to .jsx (but keep ImageWithFallback.tsx)
    content = content.replace(
      /from ['"](.+?)\.tsx['"]/g,
      (match, p1) => {
        if (p1.includes('ImageWithFallback')) {
          return match; // Keep ImageWithFallback.tsx
        }
        return `from '${p1}.jsx'`;
      }
    );
    
    // Update dynamic imports
    content = content.replace(
      /import\(['"](.+?)\.tsx['"]\)/g,
      (match, p1) => {
        if (p1.includes('ImageWithFallback')) {
          return match;
        }
        return `import('${p1}.jsx')`;
      }
    );
    
    if (content !== originalContent) {
      fs.writeFileSync(filePath, content, 'utf8');
      console.log(`✓ Updated imports in: ${filePath}`);
    }
  } catch (error) {
    console.error(`✗ Error updating ${filePath}:`, error.message);
  }
}

// Main execution
console.log('🔄 Starting TSX to JSX conversion...\n');

// Step 1: Get all .tsx files
const srcDir = path.join(__dirname, 'src');
const tsxFiles = getAllTsxFiles(srcDir);

console.log(`Found ${tsxFiles.length} .tsx files to convert\n`);

// Step 2: Rename all .tsx files to .jsx
console.log('📝 Renaming files...\n');
tsxFiles.forEach(renameFile);

// Step 3: Update entrypoint
console.log('\n📝 Updating entrypoint...\n');
const entrypointPath = path.join(__dirname, '__figma__entrypoint__.ts');
updateFileImports(entrypointPath);

// Step 4: Update all .jsx, .js, and .ts files to fix imports
console.log('\n📝 Updating imports in all files...\n');
function getAllJsFiles(dir, fileList = []) {
  const files = fs.readdirSync(dir);
  
  files.forEach(file => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    
    if (stat.isDirectory() && file !== 'node_modules') {
      getAllJsFiles(filePath, fileList);
    } else if (file.match(/\.(jsx|js|ts)$/)) {
      fileList.push(filePath);
    }
  });
  
  return fileList;
}

const allJsFiles = getAllJsFiles(__dirname);
allJsFiles.forEach(updateFileImports);

console.log('\n✅ Conversion complete!');
console.log('\nNote: ImageWithFallback.tsx was kept as .tsx (protected file)');

File: convert_tsx_to_jsx.py
#!/usr/bin/env python3

import os
import re
from pathlib import Path

# Protected files
PROTECTED_FILES = {'ImageWithFallback.tsx'}

def find_tsx_files(root_dir):
    """Find all .tsx files except protected ones"""
    tsx_files = []
    for root, dirs, files in os.walk(root_dir):
        for file in files:
            if file.endswith('.tsx') and file not in PROTECTED_FILES:
                tsx_files.append(os.path.join(root, file))
    return tsx_files

def rename_tsx_to_jsx(file_path):
    """Rename a .tsx file to .jsx"""
    new_path = file_path.replace('.tsx', '.jsx')
    try:
        os.rename(file_path, new_path)
        print(f"✓ Renamed: {file_path} -> {new_path}")
        return True
    except Exception as e:
        print(f"✗ Error renaming {file_path}: {e}")
        return False

def update_imports_in_file(file_path):
    """Update imports from .tsx to .jsx in a file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        
        # Replace .tsx with .jsx in imports, but not for ImageWithFallback
        # Match: from "./something.tsx" or from '../something.tsx' 
        def replace_import(match):
            full_match = match.group(0)
            if 'ImageWithFallback' in full_match:
                return full_match
            return full_match.replace('.tsx', '.jsx')
        
        # Update static imports
        content = re.sub(
            r'from\s+["\']([^"\']+\.tsx)["\']',
            replace_import,
            content
        )
        
        # Update dynamic imports
        content = re.sub(
            r'import\(["\']([^"\']+\.tsx)["\']\)',
            replace_import,
            content
        )
        
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"✓ Updated imports: {file_path}")
            return True
        return False
    except Exception as e:
        print(f"✗ Error updating {file_path}: {e}")
        return False

def main():
    print("🔄 Starting TSX to JSX conversion...\n")
    
    # Get the project root directory
    script_dir = Path(__file__).parent
    src_dir = script_dir / 'src'
    
    if not src_dir.exists():
        print(f"✗ Source directory not found: {src_dir}")
        return
    
    # Step 1: Find all .tsx files
    print("📁 Finding .tsx files...\n")
    tsx_files = find_tsx_files(str(src_dir))
    print(f"Found {len(tsx_files)} .tsx files to convert\n")
    
    # Step 2: Rename all .tsx to .jsx
    print("📝 Renaming files...\n")
    for tsx_file in tsx_files:
        rename_tsx_to_jsx(tsx_file)
    
    # Step 3: Update entrypoint file
    print("\n📝 Updating entrypoint...\n")
    entrypoint = script_dir / '__figma__entrypoint__.ts'
    if entrypoint.exists():
        update_imports_in_file(str(entrypoint))
    
    # Step 4: Update all imports in .jsx, .js, and .ts files
    print("\n📝 Updating imports in all JavaScript/TypeScript files...\n")
    for ext in ['.jsx', '.js', '.ts']:
        for root, dirs, files in os.walk(str(src_dir)):
            for file in files:
                if file.endswith(ext):
                    file_path = os.path.join(root, file)
                    update_imports_in_file(file_path)
    
    print("\n✅ Conversion complete!")
    print("Note: ImageWithFallback.tsx was kept as .tsx (protected file)")

if __name__ == '__main__':
    main()
